sql: make index Merge function safer for retry

stevendanna opened the pull request:
Previously, the bound `key` variable was captured by the func passed
to (*kv.DB).Txn and was used as both the start of the Scan range and the
variable we updated to save the next start key.

This is a problem because the variable would be updated before we
actually knew the transaction has commited. If the func was retried,
the next scan would then start from the updated key rather than
restarting from the original location.

Now, we always Scan from the passed startKey.

Informs #76733

Release note: None

stevendanna committed:
sql: make index Merge function safer for retry

Previously, the bound `key` variable was captured by the func passed
to (*kv.DB).Txn was used as both the start of the Scan range and the
variable we updated to save the next start key.

This is a problem because the variable would be updated before we
actually knew the transaction has commited. If the func was retried,
the next scan would then start from the updated key rather than
restarting from the original location.

Now, we always Scan from the passed startKey.

Informs #76733

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/76870)
<!-- Reviewable:end -->

stevendanna committed:
sql: add test for retried index merge transaction

This tests that an index create is successful even when the one of the
merge transactions is retried.

Release note: None

stevendanna commented:
bors r=rhu713
